down-the-stack/dk_template/src/lib.rs

```diff
@@ -0,0 +1,432 @@
+        let _ = self.inner.set_high();
+    }
+
+    /// Returns `true` if the LED is in the OFF state
+    pub fn is_off(&self) -> bool {
+        self.inner.is_set_high() == Ok(true)
+    }
+
+    /// Returns `true` if the LED is in the ON state
+    pub fn is_on(&self) -> bool {
+        !self.is_off()
+    }
+
+    /// Toggles the state (on/off) of the LED
+    pub fn toggle(&mut self) {
+        if self.is_off() {
+            self.on();
+        } else {
+            self.off()
+        }
+    }
+}
+// --- Exercise ---  🔽
+/// All buttons on the board
+pub struct Buttons {
+    /// BUTTON1: pin P0.11, green LED
+    pub b_1: Button,
+    /// BUTTON2: pin P0.12, green LED
+    pub b_2: Button,
+    /// BUTTON3: pin P0.24, green LED
+    pub b_3: Button,
+    /// BUTTON4: pin P0.25, green LED
+    pub b_4: Button,
+}
+
+/// A single button
+pub struct Button {
+    inner: Pin<Input<PullUp>>,
+}
+
+impl Button {
+    /// returns true if button is pushed 
+    pub fn is_pushed(&self) -> bool {
+        self.inner.is_low() == Ok(true)
+    }
+}
+// --- Exercise --- 🔼 
+
+/// A timer for creating blocking delays
+pub struct Timer {
+    inner: hal::Timer<hal::pac::TIMER0, OneShot>,
+}
+
+
+impl Timer {
+    /// Blocks program execution for at least the specified `duration`
+    pub fn wait(&mut self, duration: Duration) {
+        defmt::trace!("blocking for {:?} ...", duration);
+
+        // 1 cycle = 1 microsecond
+        const NANOS_IN_ONE_MICRO: u32 = 1_000;
+        let subsec_micros = duration.subsec_nanos() / NANOS_IN_ONE_MICRO;
+        if subsec_micros != 0 {
+            self.inner.delay(subsec_micros);
+        }
+
+        const MICROS_IN_ONE_SEC: u32 = 1_000_000;
+        // maximum number of seconds that fit in a single `delay` call without overflowing the `u32`
+        // argument
+        const MAX_SECS: u32 = u32::MAX / MICROS_IN_ONE_SEC;
+        let mut secs = duration.as_secs();
+        while secs != 0 {
+            let cycles = if secs > MAX_SECS as u64 {
+                secs -= MAX_SECS as u64;
+                MAX_SECS * MICROS_IN_ONE_SEC
+            } else {
+                let cycles = secs as u32 * MICROS_IN_ONE_SEC;
+                secs = 0;
+                cycles
+            };
+
+            self.inner.delay(cycles)
+        }
+
+        defmt::trace!("... DONE");
+    }
+}
+
+impl ops::Deref for Timer {
+    type Target = hal::Timer<hal::pac::TIMER0, OneShot>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl ops::DerefMut for Timer {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.inner
+    }
+}
+
+/// Uarte peripheral
+pub struct Uarte {
+    inner: hal::Uarte<hal::pac::UARTE0>,
+}
+
+impl fmt::Write for Uarte {
+
+    fn write_str(&mut self, s: &str) -> fmt::Result {
+        // Copy all data into an on-stack buffer so we never try to EasyDMA from
+        // flash.
+        let buf = &mut [0; 16][..];
```

Comment:
This is sufficient here? `buf` isn't reassigned, only sliced with the index operators.

```suggestion
        let buf = [0; 16];
```